sensor: Introduce Phosense XBR818 Driver

VynDragon opened the pull request:
This Introduces a driver for the i2c interface of Phosense XBR818.
XBR818 is a 10.525Ghz Radar chip with in-built movement detection algorithm.

VynDragon committed:
sensor: Introduce Phosense XBR818 Driver

This Introduces a driver for the i2c interface of Phosense XBR818.
XBR818 is a 10.525Ghz Radar chip with builtin detection algorithm.

Signed-off-by: Camille BAUD <mail@massdriver.space>

jilaypandya reviewed (commented):
add an entry to the build_all test 
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/build_all/sensor/i2c.dtsi

jilaypandya commented on dts/bindings/sensor/phosense,xbr818.yaml:
```diff
@@ -0,0 +1,19 @@
+    type: phandle-array
+    required: true
+    description: GPIO pin connected to IO pin (IO_VAL)
+  i2c_en-gpios:
```
Nit: add a new line before :-)

VynDragon commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,480 @@
+static int xbr818_init_defaults(const struct device *dev)
+{
+	const struct xbr818_config *config = dev->config;
+	int ret = 0;
```
If a compiler doesn't optimize that out of no-optimization choices, I'd be worried using that compiler for anything complex.

jilaypandya replied:
Have a look over here https://github.com/zephyrproject-rtos/zephyr/blob/5c9b0a5a33cab280ed29bb73e854a355fcd4194b/drivers/sensor/bosch/bmp180/bmp180.c#L106

There is a reason for it.

VynDragon replied:
I changed it already, I understand it needs to support best performance and memory footprint on compilers I'd personally consider bad.

VynDragon commented:
> add an entry to the build_all test https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/build_all/sensor/i2c.dtsi

Addressed all the comments, can you confirm what i've added to i2c.dtsi is correct?

jilaypandya commented:
> (which it shouldnt do knowing the majority of vendors are not in that list and i was told it's only a copy of the linux one, so we cant add to it)

Try adding the vendor name over here
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/vendor-prefixes.txt

VynDragon commented:
> Try adding the vendor name over here https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/vendor-prefixes.txt

Done, but I am concerned it isn't fine,  though maybe I misunderstand something.

VynDragon commented:
check error now is relevant to additional sensor attributes, not sure what to do given this is a pattern that exists in other drivers, maybe I am not doing it how it's supposed to be?
![image](https://github.com/user-attachments/assets/5867fe9c-4534-4e79-b423-2a89667a1a5b)
![image](https://github.com/user-attachments/assets/28aab1a3-038b-4ee4-81cd-686eb880dced)

jilaypandya reviewed (commented):
I believe you need to refactor the files in drivers/sensor/<vendor_name>/

VynDragon commented:
> I believe you need to refactor the files in drivers/sensor/<vendor_name>/

I dont think so, looks to be only if there is more than one sensor of the vendor (eg sensor/bosch folder vs  sensor/mhz19b or sensor/hp206c for example)

VynDragon commented:
Found a couple more cases of variable spending longer time on stack than needed so i removed the init for them too.

VynDragon commented:
ended up using a workaround to remove the enum thing, it should now pass the tests.

jilaypandya commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,473 @@
+		return -EINVAL;
+	}
+
+	gpio_pin_set_dt(&config->i2c_en, 1);
```
Handle return value [gpio_pin_set_dt](https://docs.zephyrproject.org/latest/doxygen/html/group__gpio__interface.html#ga541064fb9e575c0c559c101754466fa8)
```suggestion
	(void)gpio_pin_set_dt(&config->i2c_en, 1);
```
or
```suggestion
	ret = gpio_pin_set_dt(&config->i2c_en, 1);
	if (ret) {
	    return ret;
	}
```

VynDragon replied:
Very good point, done.

jilaypandya commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,473 @@
+
+	gpio_pin_set_dt(&config->i2c_en, 1);
+
+	switch ((int)attr) {
```
Probably add a comment as to why this cast is required

VynDragon replied:
Done.

jilaypandya commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,473 @@
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/kernel.h>
```
These are already present in `xbr818.h`

```
#include <zephyr/drivers/sensor.h>
#include <zephyr/drivers/gpio.h>
#include <zephyr/kernel.h>
```

VynDragon replied:
Done.

rruuaanng commented on dts/bindings/sensor/phosense,xbr818.yaml:
```diff
@@ -0,0 +1,20 @@
+    required: true
+    description: GPIO pin connected to IO pin (IO_VAL)
+
+  i2c_en-gpios:
```
cc @decsny Look this, I think we shouldn't use two styles of delimiters, WDYT?
```suggestion
  i2c-en-gpios:
```

VynDragon replied:
It's not delimiting anything here, it's the name for this pin (i2c_en, iic_en or iien depending on the specific board)

jeppenodgaard replied:
DTS specification v0.4 section 2.2.4 allows `_` in the property name.

However other devices seems common to use `-` in the property name. See:
`grep -Pr '.*-.*-gpios:' dts/bindings/` vs `grep -Pr '.*_.*-gpios:' dts/bindings/`
Or in Linux source:
`grep -Pr '.*-.*-gpios:' Documentation/devicetree/bindings/` vs `grep -Pr '.*_.*-gpios:' Documentation/devicetree/bindings/`

I do not like mixing delimiters either but I guess personal preference cannot be used to request changes.

rruuaanng replied:
We should follow the Linux style as much as possible. We are currently migrating all property names to hyphens unless it is used upstream.

VynDragon replied:
@rruuaanng As previously mentioned, It's not a delimiter. 
![image](https://github.com/user-attachments/assets/973fe95f-16d4-45a5-b05c-0b13730deeb6)

rruuaanng replied:
Oh, perhaps we can tolerate it for now. What I actually want to express is that it's generally encouraged to use hyphens, even though the manual uses underscores. We still need to stick with hyphens because they are typically preferred in bindings over underscores.

VynDragon replied:
I asked on the hyphen RFC, i will see what people say there about this case.

VynDragon replied:
Should be done if build passes.

VynDragon commented on tests/drivers/build_all/sensor/i2c.dtsi:
I dont think I will, it is a different area but it's part of adding this peripheral so i think it's better to keep it like this and so keep all the files related to it in one commit.

kartben commented on dts/bindings/sensor/phosense,xbr818.yaml:
```diff
@@ -0,0 +1,20 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    Phosense XBR818 I2C-capable 10Ghz Radar Sensor
```
nit: use proper convention for International System of Units

```suggestion
    Phosense XBR818 I2C-capable 10 Ghz Radar Sensor
```

VynDragon replied:
Done.

kartben commented on drivers/sensor/xbr818/xbr818.h:
```diff
@@ -0,0 +1,102 @@
+ * 0x0: t3_int_irq, 0x9: io_value_out, 0xa: io_value_out inverted, 0xf: GPIO
+ */
+#define XBR818_PIN_SETTINGS        0x23
+/* [0]: ADC1 is configuration for VCO trimming. 0: enable, 1: disable
```
```suggestion
/* [0]: ADC1 is configured for VCO trimming. 0: enable, 1: disable
```

kartben commented on drivers/sensor/xbr818/xbr818.h:
```diff
@@ -0,0 +1,102 @@
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
+
+#define SENSOR_ATTR_XBR818_DELAY_TIME  (SENSOR_ATTR_PRIV_START)
+#define SENSOR_ATTR_XBR818_LOCK_TIME   (SENSOR_ATTR_PRIV_START + 1)
+#define SENSOR_ATTR_XBR818_NOISE_FLOOR (SENSOR_ATTR_PRIV_START + 2)
+#define SENSOR_ATTR_XBR818_RF_POWER    (SENSOR_ATTR_PRIV_START + 3)
```
Those need to be in a *public* header file, which this file isn't, if you want people to be able to use this :) i.e. they need to be moved to a `include/zephyr/drivers/sensor/xbr818.h` header. 
Also please make sure this has proper Doxygen comments. See e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/sensor/tmag5273.h#L36-L49

VynDragon replied:
Done.

kartben replied:
> Also please make sure this has proper Doxygen comments.

Can you please take another pass, as Doxygen comments are missing (see, a good reason for not closing conversations after all 😉 ).
https://builds.zephyrproject.io/zephyr/pr/81013/docs/doxygen/html/xbr818_8h.html#a2f524848967d71fe6c7557355a064bec
you want something like tmag5273.h that I pointed you to, where the enum and all its fields have proper doxygen comments. (

VynDragon replied:
Then I have not properly identified what constitute a proper doxygen comment, what is it? the /** instead of /* at the start?

VynDragon replied:
Ok, figured it out.

VynDragon commented on dts/bindings/sensor/phosense,xbr818.yaml:
```diff
@@ -0,0 +1,20 @@
+
+  i2c-en-gpios:
+    type: phandle-array
+    required: true
```
It makes sense, I didnt consider the possibility of always on, only always off where my thought was 'no driver needed in that case'.

VynDragon replied:
Done.

VynDragon commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -768,6 +768,13 @@ test_i2c_tsl2540: tsl2540@71 {
 	int-gpios = <&test_gpio 0 0>;
 };
 
+test_i2c_xbr818: xrb8180@71 {
```
Isnt it supposed to match the i2c address? Yes it collides with tsl2540.

kartben replied:
Ah I see why you got confused. Not it doesn't have to be the same -- it just needs to be whatever, so that the driver can be tested/compiled in CI, so please just add at the end like I stated in the first comment and should be good!

kartben replied:
> ... so please just add at the end like I stated in the first comment and should be good!

which you have - perfect :)

VynDragon replied:
:+1: Done.

VynDragon commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,499 @@
+	ret = gpio_pin_set_dt(&config->i2c_en, 1);
+	if (ret != 0) {
+		LOG_ERR("%s: could not set i2c_en pin", dev->name);
+	}
```
Done (?).

kartben replied:
answering to your question mark with [this](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers:~:text=Unless%20they%20applied%20the%20reviewer%E2%80%99s%20recommendation%20exactly%2C%20authors%20must%20not%20resolve%20and%20hide%20comments) from the Contribution guidelines :)

Best way is to basically leave the conversation "unresolved" so that the person who request/suggested the change can go back and provide further feedback / eventually resolve the comment. This is also a good way for reviewer to not have to keep mental notes of all the things they may have asked in the past and that they need to go and double check, as it's really hard to do once all conversations have already been resolved and are collasped in the Github UI

kartben replied:
so... unresolving until I have another look :) and even if the conversation doesn't get "resolved" in Github, that deosn't prevent reviewer to eventually give a +1 anyway

VynDragon replied:
Alright, I'll leave the conversations open in the future.

VynDragon commented:
Rebased on latest main.

yperess commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,533 @@
+	if (ret < 0) {
+		return ret;
+	}
+	data->value = (ret == 1 ? true : false);
```
nit, why not just `data->value = rent == 1;`?

VynDragon replied:
I do prefer the ternary form.

yperess commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,533 @@
+	const struct xbr818_config *config = dev->config;
+	int ret;
+
+	ret = i2c_reg_write_byte_dt(&config->i2c, XBR818_IO_ACTIVE_VALUE_REG, 0x03);
```
nit: This is a lot of branches. Generally if one fails we can't recover anyway. I usually like to do:

```
int ret = 0;

ret |= i2c_*(...);
__ASSERT(ret == 0, "Some error message);

ret |= i2c_*(...);
__ASSERT(ret == 0, "Some error message);

return ret;
```

This way when I build a debug build I enable the asserts and I can see exactly when something fails, but in production I just care if the function failed and the asserts turn into no-ops.

VynDragon replied:
This is nice, done.

yperess reviewed (dismissed):
Only nits, I'm OK if you choose to ignore them. Seems fine to me.

VynDragon commented:
Rebased on latest main.

yperess commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,499 @@
+			return -EINVAL;
+		}
+		tmp = val->val1 & 0xFF;
+		ret = i2c_reg_write_byte_dt(&config->i2c, XBR818_THRESHOLD_1, tmp);
+		if (ret != 0) {
+			break;
+		}
+		tmp = (val->val1 & 0xFF00) >> 8;
+		ret = i2c_reg_write_byte_dt(&config->i2c, XBR818_THRESHOLD_2, tmp);
```
`XBR818_THRESHOLD_1` and `XBR818_THRESHOLD_2` are sequential registers, why not use `i2c_write_dt` and read both bytes in one shot? There's a big performance difference between `i2c_write_dt` vs 2 calls to `i2c_reg_write_byte_dt`. Same applies below.

VynDragon replied:
I am also seeing a mix of doing that (each register byte written one by one) vs burst writes of a register serie in existing code, say bme280 vs mcp9600. I dont think It's worth bothering for the performance impact (if the settings here are set more than once i'd be surprised),  so I dont think ill change it, Is this ok with you?

yperess replied:
@MaureenHelm is the lead on sensors, I'll let her make the call but I've seen her request the same change on other drivers. I generally prefer to keep traffic to a minimum on buses. I've seen some boards where a lot of sensors are on the same bus and the extra transactions slow down all the other drivers.

yperess reviewed (changes requested):
One more nit that I missed on the first pass, should be simple enough to update, then I think the driver looks ready.

mbolivar commented:
No objections to the vendor-prefixes.txt change. I don't see anything else for me to review here.

yperess commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,450 @@
+			break;
+		}
+		val->val1 = tmp[0] & 0xFF;
+		val->val1 |= tmp[1] << 8;
```
Can you take a second look here. I think that all these `|=` assignment don't do anything. Since `tmp` is `uint8_t` isn't `tmp[1] << 8` just 0? I think you need to cast `(uint32_t)tmp[1]` before doing the shift.

VynDragon replied:
I have tested this functionality before pushing. It works because of [integer promotions](https://stackoverflow.com/questions/46073295/implicit-type-promotion-rules). Do you want it explicited?

kartben replied:
I think it needs to be explicit, yes (MISRA rule 10.6)

VynDragon replied:
Done.

VynDragon commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,450 @@
+	/* this cast is necessary due to the additional attributes not being part of the
+	 * initial enum and causing a build error without it.
+	 */
+	switch ((int)attr) {
```
This does not in the other direction this time (SENSOR_ATTR_SAMPLING_FREQUENCY case value '0' not in enumerated type 'enum sensor_attribute_xbr818'

kartben replied:
Eh, right 😐
Guess this pattern might be better then https://github.com/zephyrproject-rtos/zephyr/blob/94cd7865a99d4c19a7e33dc4cd48b9bb9d70ac42/drivers/sensor/vishay/veml7700/veml7700.c#L412. Definitely more than one driver that's probably non conforming in-tree though...

kartben replied:
Eh, right 😐
Guess this pattern might be better then https://github.com/zephyrproject-rtos/zephyr/blob/94cd7865a99d4c19a7e33dc4cd48b9bb9d70ac42/drivers/sensor/vishay/veml7700/veml7700.c#L412. Definitely more than one driver that's probably non conforming in-tree though...

kartben replied:
Eh, right 😐
Guess this pattern might be better then https://github.com/zephyrproject-rtos/zephyr/blob/94cd7865a99d4c19a7e33dc4cd48b9bb9d70ac42/drivers/sensor/vishay/veml7700/veml7700.c#L412. Definitely more than one driver that's probably non conforming in-tree though...

VynDragon replied:
Done.

VynDragon commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -1238,3 +1238,10 @@ test_i2c_tmp435: tmp435@a9 {
 	beta-compensation = <0x0f>;
 };
+
+test_i2c_xbr818: xrb818@aa {
```
Done.

yperess commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,434 @@
+	}
+
+	if (attr == SENSOR_ATTR_LOWER_THRESH) {
+		if (val->val1 > 0xFFFF) {
```
Just noticed this, `val1` can also be negative, so you should add `|| val->val1 < 0`

VynDragon replied:
Done.

yperess commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,434 @@
+		tmp[1] = (val->val1 & 0xFF00) >> 8;
+		ret = i2c_burst_write_dt(&config->i2c, XBR818_THRESHOLD_1, tmp, 2);
+	} else if ((enum sensor_attribute_xbr818)attr == SENSOR_ATTR_XBR818_NOISE_FLOOR) {
+		if (val->val1 > 0xFFFF) {
```
Same here

VynDragon replied:
Done.

yperess commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,434 @@
+		return -ENODEV;
+	}
+
+	ret = xbr818_disable_i2c(dev);
+	if (ret != 0) {
+		return ret;
+	}
+
+	return ret;
```
Can just `return xbr818_disable_i2c(dev);` here

VynDragon replied:
Done.

rruuaanng reviewed (approved):
Thank you for taking my advice!

kartben commented:
@MaureenHelm can you have a look please? PR has gone through quite a few iterations so I'm hopeful it's good to go :)

MaureenHelm commented on drivers/sensor/xbr818/xbr818.c:
```diff
@@ -0,0 +1,434 @@
+	return ret;
+}
+
+static const struct sensor_driver_api xbr818_api = {
```
```suggestion
static DEVICE_API(sensor, xbr818_api) = {
```

VynDragon replied:
Should I submit a fix commit for this given this was merged before i had the time to update it?

VynDragon replied:
It is at https://github.com/zephyrproject-rtos/zephyr/pull/85762
